compiler/qsc_eval/src/lib.rs

```diff
@@ -1086,6 +1093,67 @@ impl State {
         }
     }
 
+    #[allow(clippy::too_many_arguments)]
+    fn eval_intrinsic(
+        &mut self,
+        env: &mut Env,
+        callee_id: StoreItemId,
+        functor: FunctorApp,
+        callee: &fir::CallableDecl,
+        sim: &mut impl Backend<ResultType = impl Into<val::Result>>,
+        callee_span: PackageSpan,
+        arg: Value,
+        arg_span: PackageSpan,
+        out: &mut impl Receiver,
+    ) -> Result<(), Error> {
+        self.push_frame(Vec::new().into(), callee_id, functor);
+        self.increment_call_count(callee_id, functor);
+        let name = &callee.name.name;
+        let val = match name.as_ref() {
+            "__quantum__rt__qubit_allocate" => {
+                let q = Rc::new(Qubit(sim.qubit_allocate()));
+                env.track_qubit(Rc::clone(&q));
```

Comment:
I'm seeing this in a couple of places in this file - why not just call the method the "normal" way?

```suggestion
                env.track_qubit(q.clone());
```

Reply:
This one is a Rust convention, though one we admittedly don't follow uniformly. Explicitly using the function syntax makes it easy to distinguish between shallow copy with a reference increment and deep copy of the underlying data. This is described in the Rust Book chapter on reference counting (see [here](https://doc.rust-lang.org/book/ch15-04-rc.html#using-rct-to-share-data), about halfway down the page):

> We could have called `a.clone()` rather than `Rc::clone(&a)`, but Rust’s convention is to use `Rc::clone` in this case. The implementation of `Rc::clone` doesn’t make a deep copy of all the data like most types’ implementations of clone do. The call to `Rc::clone` only increments the reference count, which doesn’t take much time. Deep copies of data can take a lot of time. By using `Rc::clone` for reference counting, we can visually distinguish between the deep-copy kinds of clones and the kinds of clones that increase the reference count. When looking for performance problems in the code, we only need to consider the deep-copy clones and can disregard calls to `Rc::clone`.

[There's even a clippy lint for it](https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr), though it's allow by default. I suppose if we really wanted to be consistent about this and make future investigations into deep copies easier to do we would turn that on in our repo settings. Turns out we use `Rc::clone` in 35 places (including the new ones in this PR) but turning on the lint shows we have 169 places where we call `.clone()` on ref-counted types (including three spots I introduced in this PR, whoops). I'll at least switch those three spots to `Weak::clone` for consistency in this PR, but I think it would be up for discussion if we want to be stricter about this pattern across the code base.

Reply:
Thanks, TIL! This makes sense to me.